File: submit/prj1-sol/py/tl_parser.py
#!/usr/bin/env python3

import re
import sys
from collections import namedtuple
import json

from tl_lexer import scan_stdin

"""
program                      # [ (def|expr)* ]
  : def program
  | expr program
  | // empty
  ;

def
  : 'def' id '(' formals ')' expr   # ast: DEF(id, formals, expr)
  ;

formals
  : id ? ( ',' id )*   # ast: FORMALS(id*)
  ;

expr
  : condExpr
  ;
condExpr
  : relExpr ( '?' relExpr ':' condExpr ) ?
  ;
relExpr
  : addExpr ( ( '<' | '<=' | '>' | '>=' | '==' | '!=' ) addExpr ) ?
  ;
addExpr
  : multExpr ( ( '+' | '-' ) multExpr ) *
  ;
multExpr
  : primaryExpr ( ( '*' | '/' ) primaryExpr ) *
  ;
primaryExpr
  : '-' primaryExpr
  | '(' expr ')'
  | INT
  | ID
  | ID '(' actuals ')'
  ;

actuals
  : expr ? ( ',' expr )*     # ast ACTUALS(expr*)
  ;
"""

def parse():

    def check(kind): return lookahead.kind == kind
    def match(kind):
        nonlocal lookahead
        if (lookahead.kind == kind):
            lookahead = nextToken()
        else:
            print(f'expecting {kind} at {lookahead.lexeme}',
                  file=sys.stderr)
            sys.exit(1)
    def nextToken():
        nonlocal index
        tok = tokens[index]
        index += 1
        return tok

    def program():
        asts = []
        while (not check('<EOF>')):
            ast = defn() if (check('DEF')) else expr()
            asts.append(ast)
        return asts

    def defn():
        match('DEF')
        id = lookahead.lexeme
        match('ID')
        match('(')
        f = formals()
        match(')')
        e = expr()
        idAst = Ast('ID')
        idAst['id'] = id
        return Ast('DEF', idAst, f, e)

    def formals():
        list = []
        if check('ID'):
            idAst = Ast('ID')
            idAst['id'] = lookahead.lexeme
            list.append(idAst)
            match('ID')
            while check(','):
                match(',')
                idAst = Ast('ID')
                idAst['id'] = lookahead.lexeme
                list.append(idAst)
                match('ID')
        return Ast('FORMALS', *list)

    def expr():
        return condExpr()

    def condExpr():
        e = relExpr()
        if (check('?')):
            match('?')
            e1 = relExpr()
            match(':')
            e2 = condExpr()
            e = Ast('?:', e, e1, e2)
        return e

    def relExpr():
        e = addExpr()
        while (check('<') or check('<=') or check('>') or check('>=') or
               check('==') or check('!=')):
            relOp = lookahead.kind
            match(relOp)
            e1 = addExpr()
            e = Ast(relOp, e, e1)
        return e

    def addExpr():
        e = multExpr()
        while (check('+') or check('-')):
            op = lookahead.kind
            match(op)
            e1 = multExpr()
            e = Ast(op, e, e1)
        return e

    def multExpr():
        e = primaryExpr()
        while (check('*') or check('/')):
            op = lookahead.kind
            match(op)
            e1 = primaryExpr()
            e = Ast(op, e, e1)
        return e

    def primaryExpr():
        if check('-'):
            match('-')
            return Ast('-', primaryExpr())
        if check('('):
            match('(')
            e = expr()
            match(')')
            return e
        if check('INT'):
            value = int(lookahead.lexeme)
            match('INT')
            ast = Ast('INT')
            ast['value'] = value
            return ast
        id = lookahead.lexeme
        match('ID')
        ast = Ast('ID')
        ast['id'] = id
        if not check('('):
            return ast
        match('(')
        exprs = actuals()
        match(')')
        return Ast('APP', ast, exprs)

    def actuals():
        list = []
        if not check(')'):
            e = expr()
            list.append(e)
            while check(','):
                match(',')
                e = expr()
                list.append(e)
        return Ast('ACTUALS', *list)

    #begin parse()
    tokens = scan_stdin()
    index = 0
    lookahead = nextToken()
    value = program()
    if (not check('<EOF>')):
        print(f'expecting <EOF>, got {lookahead.lexeme}', file=sys.stderr)
        sys.exit(1)
    return value

def main():
    asts = parse()
    print(json.dumps(asts, separators=(',', ':'))) #no whitespace

#use a dict so that we can add attributes dynamically
def Ast(tag, *kids):
    return { 'tag': tag, 'kids': kids }

if __name__ == "__main__":
    main()

File: submit/prj1-sol/js/tl-parser.mjs
#!/usr/bin/env node

import scan from './tl-lexer.mjs';

import fs from 'fs';
import Path from 'path';

/**
program                      # [ (def|expr)* ]
  : def program
  | expr program
  | // empty 
  ;

def
  : 'def' id '(' formals ')' expr   # ast: DEF(id, formals, expr)
  ;

formals
  : id ? ( ',' id )*   # ast: FORMALS(id*)
  ;

expr
  : condExpr
  ;
condExpr
  : relExpr ( '?' relExpr ':' condExpr ) ?
  ;
relExpr
  : addExpr ( ( '<' | '<=' | '>' | '>=' | '==' | '!=' ) addExpr ) ?
  ;
addExpr
  : multExpr ( ( '+' | '-' ) multExpr ) *
  ;
multExpr
  : primaryExpr ( ( '*' | '/' ) primaryExpr ) *
  ;
primaryExpr
  : '-' primaryExpr
  | '(' expr ')' 
  | INT
  | ID	
  | ID '(' actuals ')' 
  ;

actuals
  : expr ? ( ',' expr )*     # ast ACTUALS(expr*)
  ;
 */

class TlParser {
  constructor(tokens) {
    this._tokens = tokens;
    this._index = -1;
    this.lookahead = undefined;
    this._nextToken();
  }

  //wrapper used for crude  error recovery
  parse() {
    try {
      let result = this.program();
      if (!this.check('<EOF>')) {
	const msg = `expecting end-of-file at "${this.lookahead.lexeme}"`;
	throw new SyntaxError(msg);
      }
      return result;
    }
    catch (err) {
      return err;
    }
  }

  check(kind) {
    return this.lookahead.kind === kind;
  }
  
  match(kind) {
    if (this.check(kind)) {
      this._nextToken();
    }
    else {
      const msg = `expecting ${kind} at "${this.lookahead.lexeme}"`;
      throw new SyntaxError(msg);
    }
  }

  _nextToken() {
    if (this._index < this._tokens.length - 1) {
      this._index++;
      this.lookahead  = this._tokens[this._index];
    }
    else {
      this.lookahead = null;
    }
  }
  

  program() {
    const asts = [];
    while (!this.check('<EOF>')) {
      const ast = (this.check('DEF')) ? this.def() : this.expr();
      asts.push(ast);
    }
    return asts;
  }

  def() {
    this.match('DEF');
    const tok = this.lookahead;
    this.match('ID');
    const idAst = new Ast('ID');
    idAst.id = tok.lexeme;
    this.match('(');
    const formalsAst = this.formals();
    this.match(')');
    const exprAst = this.expr();
    return new Ast('DEF', idAst, formalsAst, exprAst);
  }

  formals() {
    const idAsts = [];
    if (this.check('ID')) {
      const id = this.lookahead.lexeme;
      this.match('ID');
      const idAst = new Ast('ID');
      idAst.id = id;
      idAsts.push(idAst);
      while (this.check(',')) {
	this.match(',');
	const id = this.lookahead.lexeme;
	this.match('ID');
	const idAst = new Ast('ID');
	idAst.id = id;
	idAsts.push(idAst);
      }
    }
    return new Ast('FORMALS', ...idAsts);
  }

  expr() {
    return this.condExpr();
  }

  condExpr() {
    let expr = this.relExpr();
    if (this.check('?')) {
      this.match('?');
      const thenExpr = this.relExpr();
      this.match(':');
      const elseExpr = this.condExpr();
      expr = new Ast('?:', expr, thenExpr, elseExpr);
    }
    return expr;
  }

  relExpr() {
    let exprAst = this.addExpr();
    if (this.check('<') || this.check('<=') ||
	this.check('>') || this.check('>=') ||
	this.check('==') || this.check('!=')) {
      const relOp = this.lookahead.kind;
      this.match(relOp);
      const addExprAst = this.addExpr();
      exprAst = new Ast(relOp, exprAst, addExprAst);
    }
    return exprAst;
  }

  addExpr() {
    let exprAst = this.multExpr();
    while (this.check('+') || this.check('-')) {
      const addOp = this.lookahead.kind;
      this.match(addOp);
      const multExprAst = this.multExpr();
      exprAst = new Ast(addOp, exprAst, multExprAst);
    }
    return exprAst
  }

  multExpr() {
    let exprAst = this.primaryExpr();
    while (this.check('*') || this.check('/')) {
      const multOp = this.lookahead.kind;
      this.match(multOp);
      const primaryExprAst = this.primaryExpr();
      exprAst = new Ast(multOp, exprAst, primaryExprAst);
    }
    return exprAst
  }


  primaryExpr() {
    if (this.check('-')) {
      this.match('-');
      const p = this.primaryExpr();
      return new Ast('-', p);
    }
    else if (this.lookahead.kind == 'INT') {
      const lexeme = this.lookahead.lexeme;
      this.match('INT');
      const ast = new Ast('INT');
      ast.value = Number.parseInt(lexeme);
      return ast;
    }
    else if (this.check('(')) {
      this.match('(');
      const e = this.expr();
      this.match(')');
      return e;
    }
    else {
      const id = this.lookahead.lexeme;
      this.match('ID');
      const idAst = new Ast('ID');
      idAst.id = id;
      if (this.check('(')) {
	this.match('(');
	const actuals = this.actuals();
	this.match(')');
	return new Ast('APP', idAst, actuals);
      }
      else {
	return idAst;
      }
    }
  }

  actuals() {
    const exprAsts = [];
    if (!this.check(')')) {
      exprAsts.push(this.expr());
      while (this.check(',')) {
	this.match(',');
	exprAsts.push(this.expr());
      }
    }
    return new Ast('ACTUALS', ...exprAsts);
  }



} //AstParser

class Ast {
  constructor(tag, ...kids) {
    this.tag = tag;
    this.kids = kids;
  }
}

const CHAR_SET = 'utf8'
function main() {
  const text = fs.readFileSync(0, CHAR_SET);
  const tokens = scan(text);
  const parser = new TlParser(tokens);
  const result = parser.parse();
  if (result instanceof Error) {
    console.error(result.message);
  }
  else {
    console.log(JSON.stringify(result));
  }
}

main();

